### PR TITLE
Cherrypick `2fc9feb` into r1.6

### PR DESCRIPTION
**Pull request by jysohn23**

### CODE REVIEW

File: torch_xla/core/xla_model.py [hunk 1]
```diff
@@ -376,14 +457,73 @@ def _fetch_gradients(optimizer):
...
+  token = getattr(devctx, 'all_reduce_token', None)
   if token is None:
     token = torch_xla._XLAC._xla_create_token()
-    _TLS.all_reduce_token = token
-  return token
+    devctx.all_reduce_token = token
+  return token, devctx
+
+
+def _torch_all_reduce(reduce_type, inputs, group=None):
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py:468]: You can remove from here ... (sea-of-devices)

File: torch_xla/core/xla_model.py [hunk 2]
```diff
@@ -376,14 +457,73 @@ def _fetch_gradients(optimizer):
...
+      inputs[i].copy_(results[i])
+      if scale is not None:
+        inputs[i].mul_(scale)
+  else:
+    for tensor in inputs:
+      tensor.zero_()
+  if cctx.requires_intercore_reduce:
+    token, devctx = _get_all_reduce_token()
+    devctx.all_reduce_token = torch_xla._XLAC._xla_all_reduce_inplace(
+        REDUCE_SUM, inputs, token, 1.0, [])
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py:523]: ... to here.

File: torch_xla/core/xla_model.py [hunk 3]
```diff
@@ -376,14 +457,73 @@ def _fetch_gradients(optimizer):
...
 
 
-def all_reduce(reduce_type, inputs, scale=1.0, groups=None):
+def all_reduce(reduce_type, inputs, scale=1.0, groups=None, cctx=None):
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py:526]: Here keep the original function, and look how the other collective APIs do with:
```Python
token, devctx = _get_all_reduce_token()
```

File: torch_xla/core/xla_model.py [hunk 4]
```diff
@@ -577,10 +741,12 @@ def reduce_gradients(optimizer, groups=None):
         the `[4, 5, 6, 7]` replicas. If `None` there will be only one group with
         all the replicas in it.
   """
-  count = torch_xla._XLAC._xla_get_replication_devices_count()
+  cctx = CollectiveContext()
```
Comment (dlibenzi) [inline torch_xla/core/xla_model.py:744]: No changes in this API.

File: torch_xla/core/xla_model.py [hunk 5]
```diff
@@ -15,14 +15,100 @@
...
+
 REDUCE_SUM = 'sum'
 REDUCE_MUL = 'mul'
 REDUCE_AND = 'and'
 REDUCE_OR = 'or'
 REDUCE_MIN = 'min'
 REDUCE_MAX = 'max'
 
-_TLS = threading.local()
+_TORCH_DIST_GROUPS = dict()
```
[the remark this answers is not in the document]
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Done.

File: torch_xla/core/xla_model.py [hunk 6]
```diff
@@ -15,14 +15,100 @@
...
+    device = str(device)
+  with _DEVICE_CONTEXTS_LOCK:
+    devctx = _DEVICE_CONTEXTS.get(device, None)
+    if devctx is None:
+      devctx = DeviceContext(device)
+      _DEVICE_CONTEXTS[device] = devctx
+    return devctx
+
+
+class CollectiveContext(object):
```
[the remark this answers is not in the document]
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Thanks, done.

File: torch_xla/core/xla_model.py [hunk 7]
```diff
@@ -15,14 +15,100 @@
...
+
+
+def _make_interhost_group(replica_devcount, world_size):
+  # Every host in a sea-of-devices case handles replica_devcount devices.
+  # The replica device index 0 of each host does the inter-host replication
+  # using torch.distributed.
+  # The XLA CPU is a special case where there is one process per XLA CPU device,
+  # which is also a virtual host within a physical host.
+  ranks = tuple(range(0, world_size, replica_devcount))
+  return _get_torch_dist_group(ranks), ranks
```
[the remark this answers is not in the document]
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Done.

File: torch_xla/core/xla_model.py (hunk 1 above)
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Done.

File: torch_xla/core/xla_model.py (hunk 2 above)
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Ditto, done.

File: torch_xla/core/xla_model.py (hunk 3 above)
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Done, running CI to check if all is OK.

File: torch_xla/core/xla_model.py (hunk 4 above)
Comment (jysohn23) [reply torch_xla/core/xla_model.py]: Done.

**Pull request merged**